src/emqx_connection.erl

```diff
@@ -594,18 +597,34 @@ serialize_and_inc_stats_fun(#state{serialize = Serialize}) ->
 %% Send data
 
 -spec(send(iodata(), state()) -> ok).
-send(IoData, #state{transport = Transport, socket = Socket}) ->
+send(IoData, #state{transport = Transport, socket = Socket, channel = Channel}) ->
     Oct = iolist_size(IoData),
     ok = emqx_metrics:inc('bytes.sent', Oct),
     emqx_pd:inc_counter(outgoing_bytes, Oct),
-    case Transport:async_send(Socket, IoData) of
+    may_warn_congestion(Socket, Channel),
+    case Transport:async_send(Socket, IoData, [nosuspend]) of
         ok -> ok;
         Error = {error, _Reason} ->
             %% Send an inet_reply to postpone handling the error
             self() ! {inet_reply, Socket, Error},
             ok
     end.
 
+may_warn_congestion(Socket, Channel) ->
+    Congested = erlang:get(conn_congested),
```

Comment:
`IsCongestionAlarmSet = is_congestion_alarm_set(),`
two more similar functions: `set_congestion_alarm/0` and `clear_congeston_alarm/0`